### PR TITLE
feat: add paho mqtt c lib

### LINKED ISSUE #916: [request] port paho-mqtt-cpp to CCI
### Package Details
  * Package Name/Version: **paho-mqtt-cpp/3.15.3**
  * Website: **https://github.com/conan-community/conan-paho-cpp**
  * Source code: **https://github.com/eclipse/paho.mqtt.cpp**

 depends on https://github.com/conan-io/conan-center-index/pull/914

### PR DESCRIPTION
**Pull request by gocarlos**
Specify library name and version:  **paho-c/1.3.0**

based on https://github.com/conan-community/conan-paho-c

closes #916 

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    license = "EPL-1.0"
+    description = """The Eclipse Paho project provides open-source client implementations of MQTT
+    and MQTT-SN messaging protocols aimed at new, existing, and emerging applications for the Internet
+    of Things (IoT)"""
+    exports_sources = ["CMakeLists.txt", "patches/*"]
+    generators = "cmake", "cmake_find_package"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "SSL": [True, False],
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: lowercase this option

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
+
+    def requirements(self):
+        if self.options.SSL:
+            self.requires("OpenSSL/1.1.1d")
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: The package should also be lowercase: `openssl/1.1.1d`

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    name = "paho-mqtt-c"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/eclipse/paho.mqtt.c"
+    topics = ("MQTT", "IoT", "eclipse", "SSL", "paho", "C")
+    license = "EPL-1.0"
+    description = """The Eclipse Paho project provides open-source client implementations of MQTT
+    and MQTT-SN messaging protocols aimed at new, existing, and emerging applications for the Internet
+    of Things (IoT)"""
+    exports_sources = ["CMakeLists.txt", "patches/*"]
+    generators = "cmake", "cmake_find_package"
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: The `cmake_find_package` generator is not needed. `paho-mqtt-c` does not use `find_package`.

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import os
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: Not used

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions["PAHO_ENABLE_TESTING"] = False
+        cmake.definitions["PAHO_BUILD_DOCUMENTATION"] = False
+        cmake.definitions["PAHO_BUILD_SAMPLES"] = False
+        cmake.definitions["PAHO_BUILD_STATIC"] = not self.options.shared
+        cmake.definitions["PAHO_WITH_SSL"] = self.options.SSL
+        cmake.definitions["PAHO_BUILD_ASYNC"] = self.options.asynchronous
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: You can optionally add these lines to make sure the cmake script finds the correct openssl versoin and not the system installed one:
```
        if self.options.SSL:
            cmake.definitions["OPENSSL_SEARCH_PATH"] = self.deps_cpp_info["openssl"].rootpath
```

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        if self.options.SSL:
+            self.requires("OpenSSL/1.1.1d")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name.replace("-", ".") + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: add caching:
- add global `_cmake = None` at the top
- use `self._cmake` here

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        self.copy("notice.html", src=self._source_subfolder, dst="licenses")
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        if self.settings.os == "Windows":
+            if not self.options.shared:
+                self.cpp_info.libs.append("ws2_32")
```
Comment (madebr) [inline recipes/paho-mqtt-c/all/conanfile.py]: Use `self.cpp_info.system_libs`
Same applies to other libraries below.

**Comment by gocarlos** [comment]
@danimtb anything else that should be done here to get this merged?

**Comment by gocarlos** [comment]
@SSE4 deprecate https://github.com/conan-community/conan-paho-c ?

**Comment by uilianries** [comment]
Yeah, I'll do it!

**Pull request merged**